src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs

```diff
@@ -198,4 +198,101 @@ public MultiGroupBroadcastDataMessage(IReadOnlyList<string> groupList, IDictiona
+        /// <param name="payloads">The payload of the message.</param>
+        /// <param name="tracingId">The tracing Id of the message.</param>
+        public ClientInvocationMessage(string invocationId, string connectionId, string callerId, IDictionary<string, ReadOnlyMemory<byte>> payloads, ulong? tracingId = null)
+            : base(payloads, tracingId)
+        {
+            InvocationId = invocationId;
+            ConnectionId = connectionId;
+            CallerId = callerId;
+        }
+
+        /// <summary>
+        /// Gets or sets the Id of the client invocation request.
+        /// </summary>
+        public string InvocationId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the client connection Id.
+        /// </summary>
+        public string ConnectionId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the caller server Id that init the client invocation.
+        /// </summary>
+        public string CallerId { get; set; }
+    }
+
+    /// <summary>
+    /// A data message of client invocation completion result.
+    /// </summary>
+    public class ServiceCompletionMessage : MulticastDataMessage
+    {
+        /// <summary>
+        /// Initialize a new instance of <see cref="ServiceCompletionMessage"/> class.
+        /// </summary>
+        /// <param name="invocationId">The Id of the invocation that has completed.</param>
+        /// <param name="connectionId">The client connection Id that complete the invocation.</param>
+        /// <param name="callerId">The serverId that wrap the completion result.</param>
+        /// <param name="payloads">The payload of the completion result.</param>
+        /// <param name="tracingId">The tracing Id of the message.</param>
+        public ServiceCompletionMessage(string invocationId, string connectionId, string callerId, IDictionary<string, ReadOnlyMemory<byte>> payloads, ulong? tracingId = null)
+            : base(payloads, tracingId)
+        {
+            InvocationId = invocationId;
+            ConnectionId = connectionId;
+            CallerId = callerId;
+            Payloads = payloads;
+        }
+
+        /// <summary>
+        /// Initialize a new instance of <see cref="ServiceCompletionMessage"/> class with error information.
+        /// </summary>
+        /// <param name="invocationId">The Id of the invocation that has completed.</param>
+        /// <param name="connectionId">The client connection Id that complete the invocation.</param>
+        /// <param name="callerId">The serverId that wrap the completion result.</param>
+        /// <param name="error">The error information about invation failure.</param>
+        /// <param name="tracingId">The tracing Id of the message.</param>
+        public ServiceCompletionMessage(string invocationId, string connectionId, string callerId, string error, ulong? tracingId = null)
+            : base(null, tracingId)
```

Comment:
in which case is the payload null? is it for CompletionMessage.Empty? shall we always preserve the original payload for CompletionMessage?

Reply:
In error cases that won't get a `CompletionMessage` from client, that service side send with error to notify caller server.